packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -4240,7 +4253,16 @@ class _UpdateTextSelectionAction<T extends DirectionalCaretMovementIntent> exten
     }
 
     final _TextBoundary textBoundary = getTextBoundariesForIntent(intent);
-    final TextSelection textBoundarySelection = textBoundary.textEditingValue.selection;
+
+    // textBoundary selection is only updated after rebuild; If the text value is same,
+    // use selection from state which is more recent. This is necessary on macOS
+    // where alt+up sends moveBackward: and `moveToBeginningOfParagraph:` selectors at
+    // the same time.
```

Comment:
"`textBoundary`'s selection is only updated after rebuild; if the text is the same, use the selection from state, which is more recent. This is necessary on macOS where alt+up sends the `moveBackward:` and `moveToBeginningOfParagraph:` selectors at the same time."